src/main/java/org/xmpp/packet/Roster.java

```diff
@@ -164,36 +162,109 @@ public Item addItem(JID jid, String name, Ask ask, Subscription subscription,
         if (subscription == null) {
             throw new NullPointerException("Subscription cannot be null");
         }
+
+        Element item = null;
+        List<Attribute> attributes = null;
+        String bareJID = jid.toBareJID();
         Element query = element.element(new QName("query", Namespace.get("jabber:iq:roster")));
         if (query == null) {
             query = element.addElement("query", "jabber:iq:roster");
-        }
-        Element item = null;
-        for (Iterator<Element> i=query.elementIterator("item"); i.hasNext(); ) {
-            Element el = i.next();
-            if (el.attributeValue("jid").equals(jid.toString())) {
-                item = el;
+        } else {
+            // raw iteration too speed up
+            for (Element node : (List<Element>) query.elements()) {
+                if ("item".equals(node.getName())) {
+                    attributes = node.attributes();
+                    for (Attribute attr : attributes) {
+                        // there was comparing just with jid.toString, so replace with startsWith to
+                        // process both jid format (with and without /resource)
+                        if ("jid".equals(attr.getName())) {
+                            if (attr.getValue() != null && attr.getValue().startsWith(bareJID)) {
+                                item = node; //found!
+                            }
+                            break;
+                        }
+                    }
+                    if (item != null) {
+                        break;
+                    }
+                }
             }
         }
+
+        Attribute nameAttr = null;
+        Attribute askAttr = null;
+        Attribute subscriptionAttr = null;
         if (item == null) {
             item = query.addElement("item");
+            item.addAttribute("jid", bareJID);
+            // Add in groups, to speed up no need to check existence
+            if (groups != null && !groups.isEmpty()) {
+                for (String group : groups) {
+                    item.addElement("group").setText(group);
+                }
+            }
+        } else {

```

Comment:
Algorithm was refactored to do less DOM modifications in groups merge and to avoid using very costly DOM operations such as 'addAttribute' and `elementIterator`. So all of it is about to update only value of existed attribute (not whole attribute processing) and carefully merge item group changes. It looks more complicated but works much faster.